test/performance/odp_random.c

```diff
@@ -97,198 +121,145 @@ static int parse_options(int argc, char *argv[])
-	uint64_t nsec;
-	uint32_t rounds;
-	test_shm_t *shm = shm_lookup();
-
-	if (!shm) {
-		ODPH_ERR("Failed to look up shm %s\n", shm_name);
-		exit(EXIT_FAILURE);
-	}
-
-	rounds = type_rounds(shm->type);
-
-	/* One extra page for alignment. */
-	buf = (uint8_t *)malloc(options.size + page);
-
-	if (!buf) {
-		ODPH_ERR("Memory allocation failed.\n");
-		exit(EXIT_FAILURE);
-	}
-
-	/* Align to start of page. */
-	data = (uint8_t *)(((uintptr_t)buf + (page - 1)) & ~(page - 1));
-
-	odp_barrier_wait(&shm->barrier);
-	start = odp_time_local();
-
-	for (uint32_t i = 0; i < rounds; i++) {
+	for (i = 0; i < rounds; i++) {
 		uint32_t pos = 0;
 
-		while (pos < options.size) {
-			int32_t n = odp_random_data(data + pos,
-						    options.size - pos,
-						    shm->type);
+		while (pos < size) {
+			ret = odp_random_data(data + pos, size - pos, type);
 
-			if (n < 0) {
+			if (ret < 0) {
 				ODPH_ERR("odp_random_data() failed\n");
 				exit(EXIT_FAILURE);
 			}
 
-			pos += n;
+			pos += ret;
 		}
 	}
-
-	nsec = odp_time_diff_ns(odp_time_local(), start);
-	shm->nsec[odp_thread_id()] = nsec;
-	free(buf);
-
-	return 0;
 }
 
-static int test_random_test(void *p)
+static inline void pseudo_random_data_loop(uint32_t rounds, uint8_t *data, uint32_t size)
 {
-	(void)p;
-
-	uint8_t *buf, *data;
-	const unsigned long page = ODP_PAGE_SIZE;
-	odp_time_t start;
-	uint64_t nsec;
+	uint32_t i;
+	int32_t ret;
 	uint64_t seed = 0;
-	uint32_t rounds;
-	test_shm_t *shm = shm_lookup();
-
-	if (!shm) {
-		ODPH_ERR("Failed to look up shm %s\n", shm_name);
-		exit(EXIT_FAILURE);
-	}
-
-	rounds = type_rounds(shm->type);
-
-	/* One extra page for alignment. */
-	buf = (uint8_t *)malloc(options.size + page);
-
-	if (!buf) {
-		ODPH_ERR("Memory allocation failed.\n");
-		exit(EXIT_FAILURE);
-	}
-
-	/* Align to start of page. */
-	data = (uint8_t *)(((uintptr_t)buf + (page - 1)) & ~(page - 1));
-
-	odp_barrier_wait(&shm->barrier);
-	start = odp_time_local();
 
-	for (uint32_t i = 0; i < rounds; i++) {
+	for (i = 0; i < rounds; i++) {
 		uint32_t pos = 0;
 
-		while (pos < options.size) {
-			int32_t n = odp_random_test_data(data + pos,
-							 options.size - pos,
-							 &seed);
+		while (pos < size) {
+			ret = odp_random_test_data(data + pos, size - pos, &seed);
 
-			if (n < 0) {
-				ODPH_ERR("odp_random_data() failed\n");
+			if (ret < 0) {
+				ODPH_ERR("odp_random_test_data() failed\n");
 				exit(EXIT_FAILURE);
 			}
 
-			pos += n;
+			pos += ret;
 		}
 	}
+}
+
+static int test_random_perf(void *ptr)
+{
+	odp_time_t start;
+	uint64_t nsec;
+	thread_arg_t *thread_arg = ptr;
+	test_global_t *global = thread_arg->global;
+	odp_random_kind_t type = global->type;
+	int thread_idx = thread_arg->thread_idx;
+	uint8_t *data = thread_arg->data;
+	uint32_t size = options.size;
+	uint32_t rounds = global->rounds;
+
+	/* One warm up round */
+	if ((int)type == PSEUDO_RANDOM)
+		pseudo_random_data_loop(1, data, size);
+	else
+		random_data_loop(type, 1, data, size);
```

Comment:
This seems to repeat, could be in a separate function or alternatively `*_data_loop` signatures could be unified and called via a pointer.

Reply:
I think a new function is not necessary in this case, as it would save so few lines (add 6 new lines, remove 9 lines).

Reply:
OK. In general, it's a bit cumbersome to require the type check everytime there's a call to get random data.

Reply:
Fixed in v2. Moved pseudo_random_data_loop() inside random_data_loop() as it had all needed arguments already.